source/game/system/CourseMap.cc

```diff
@@ -182,6 +182,434 @@ f32 CourseMap::startTmp3() const {
+        if (id == -1) {
+            for (size_t i = 0; i < checkpoint->prevCount(); i++) {
+                MapdataCheckPoint *checkpoint_ =
+                        i < checkpoint->prevCount() ? checkpoint->prevPoint(i) : nullptr;
+                id = findRecursiveSector(pos, 1, true, *checkpoint_, checkpointCompletion, params);
+                if (id != -1) {
+                    break;
+                }
+            }
+        }
+
+    default:
+        break;
+    }
+
+    // GLOBAL SEARCH
+    // if local search fails, remove depth limit and search all "loaded" checkpoints
+
+    // Step 1: Search all next checkpoints until player or key checkpoint is found
+    if (id == -1) {
+        for (size_t i = 0; i < checkpoint->nextCount(); i++) {
+            MapdataCheckPoint *checkpoint_ =
+                    i < checkpoint->nextCount() ? checkpoint->nextPoint(i) : nullptr;
+            id = findRecursiveSector(pos, -1, false, *checkpoint_, checkpointCompletion, 0);
+            if (id != -1) {
+                break;
+            }
+        }
+        // Step 2: Search all previous checkpoints until player or key checkpoint is found
+        if (id == -1) {
+            for (size_t i = 0; i < checkpoint->prevCount(); i++) {
+                MapdataCheckPoint *checkpoint_ =
+                        i < checkpoint->prevCount() ? checkpoint->prevPoint(i) : nullptr;
+                id = findRecursiveSector(pos, -1, true, *checkpoint_, checkpointCompletion, 0);
+                if (id != -1) {
+                    break;
+                }
+            }
+        }
+    }
+
+    // ONLINE PLAYERS ONLY
+    // if both local and global searches fail, just iterate through ALL checkpoints (ignoring kcps)
+    // until player is found or last checkpoint reached
+    if (isRemote && id == -1) {
+        for (size_t i = 0; i < m_checkPoint->size(); i++) {
+            MapdataCheckPoint *checkpoint_ = getCheckPoint(i);
+            if (!checkpoint_->searched()) { // search all unsearched checkpoints
+                MapdataCheckPoint::SectorOccupancy completion =
+                        checkpoint_->checkSectorAndCheckpointCompletion(pos, checkpointCompletion);
+                checkpoint_->setSearched();
+                if (completion == MapdataCheckPoint::SectorOccupancy::InsideSector) {
+                    id = i;
+                    break;
+                }
+            }
+        }
+    }
+
+    return id;
+}
+
+s16 CourseMap::searchNextCheckpoint(const EGG::Vector3f &pos, s16 depth,
+        const MapdataCheckPoint &checkpoint, float *completion, u32 params,
+        const bool param_8) const {
+    s16 id = -1;
+    // increment depth counter unless it's -1
+    s16 depth_ = depth >= 0 ? depth + 1 : -1;
+
+    // iterate through each next checkpoint
+    for (size_t i = 0; i < checkpoint.nextCount(); i++) {
+        MapdataCheckPoint *checkpoint_ =
+                (s32)i < checkpoint.nextCount() ? checkpoint.nextPoint(i) : nullptr;
```

Comment:
The size check is redundant since we have it in our loop condition. **Apply this to the other loops in this source file, where applicable**
```suggestion
        MapdataCheckPoint *checkpoint_ = checkpoint.nextPoint(i);
```